app/models/message_period.rb

```diff
@@ -113,14 +113,31 @@ def result
     channels = @channels.empty? ?
       [] : Channel.where(identifier: @channels.split).to_a
 
+    prev_m = {}
     messages = Message.
       filter_by_channels(channels).
       filter_by_since(@since).
       filter_by_until(@until).
       order(timestamp: :asc, id: :asc).
       limit(RESULT_LIMIT + 1).
       includes(:channel, :irc_user).
-      to_a
+      map do |m|
+        if(
+            [Quit, Nick].include?(m.class) &&
+            prev_m[:type] == m[:type] &&
+            prev_m[:timestamp] == m[:timestamp] &&
+            prev_m[:irc_user_id] == m[:irc_user_id] &&
+            prev_m[:nick] == m[:nick] &&
+            prev_m[:message] == m[:message]
+        )
+          prev_m = m
+          nil
+        else
+          prev_m = m
+          m
+        end
+      end.
+      compact
```

Comment:
同じ種類のメッセージが連続していたら飛ばすという処理ですね。概ね大丈夫だとは思いますが、単純にひとつ前のメッセージと比較すると、ほぼ同時に他の種類のメッセージが届いたとき、飛ばす処理が中途半端なところで止まるかもしれません。

対策はこんな感じでしょうか？

* NICKとQUITの明示的な指定が気になったので、これらを「参加中の全チャンネルに同時配信される（broadcast）メッセージ」と考えて、モデルにその性質を書いてみました。
* 処理が長くなるので、ファイルを分割してみました。

# `app/models/message.rb`

```ruby
# @return [Boolean] 参加中の全チャンネルに同時に送られるメッセージか？
def broadcast?
  false
end

# 同じ同時配信メッセージかどうか判定する
# @param [Message] m 判定対象のメッセージ
# @return [Boolean]
def same_broadcast_message?(m)
  broadcast? &&
    m.class == self.class &&
    m.timestamp == timestamp &&
    m.irc_user_id == irc_user_id &&
    m.nick == nick &&
    m.message == message
end
```

# `app/models/{nick,quit}.rb`

```ruby
# @return [Boolean] 参加中の全チャンネルに同時に送られるメッセージか？
def broadcast?
  true
end
```

# `app/models/conversation_message.rb`

```ruby
# @return [Boolean] 参加中の全チャンネルに同時に送られるメッセージか？
def broadcast?
  false
end
```

# `app/models/message_period.rb`

## `result` メソッド内

処理が長くなるので、`messages` は変えずに別メソッドで処理するようにします（実は別issueとしてMessageとConversationMessageの抽出処理を共通化する考えがありまして、それと合わせることも狙っています）。

```ruby
# messagesは変えない
messages = Message.
  filter_by_channels(channels).
  filter_by_since(@since).
  filter_by_until(@until).
  order(timestamp: :asc, id: :asc).
  limit(RESULT_LIMIT + 1).
  includes(:channel, :irc_user).
  to_a

united_messages = unite_broadcast_messages(messages)
# 以降はunited_messagesを使う
```

## `private` 以降

```ruby
# 参加中の全チャンネルに同時に送られるメッセージ（Nick、Quit）をまとめる
# @param [Array<Message>] messages メッセージの配列
# @return [Array<Message>] まとめられたメッセージの配列
def unite_broadcast_messages(messages)
  # 種類ごとに最後のメッセージを記録するためのハッシュ
  last = {}

  united_messages = messages.map do |m|
    # 同時配信されないメッセージはそのまま残す
    next m unless m.broadcast?

    # 同じ同時配信メッセージと見られる場合は記録しない。
    # そうでなければ残す。
    m_new = last[m.class]&.same_broadcast_message?(m) ? nil : m

    # 調べたメッセージを最後のメッセージとして記録する
    last[m.class] = m

    m_new
  end

  united_messages.compact
end
```

Reply:
messages を「変えない」で「以降はunited_messages」を使うのは、united_broadcast_messages の戻り値を messages に格納するのではいけないのでしょうか？
その方が保持する変数(メモリ量)が減ると思うので、何か支障がなければそういう実装をしたいです。

Reply:
それで大丈夫です:+1: